File: days/day_10/solutions/solution.java
import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.util.regex.*;

class Solution {
    static class ParseResult {
        List<Boolean> targetPattern = new ArrayList<>();
        List<List<Integer>> buttons = new ArrayList<>();
        List<Integer> joltages = new ArrayList<>();
    }
    
    static ParseResult parseLine(String line) {
        ParseResult result = new ParseResult();
        
        // Extract pattern: [.##.]
        Pattern patternRe = Pattern.compile("\\[([.#]+)\\]");
        Matcher patternMatch = patternRe.matcher(line);
        if (patternMatch.find()) {
            String patternStr = patternMatch.group(1);
            for (char c : patternStr.toCharArray()) {
                result.targetPattern.add(c == '#');
            }
        }
        
        // Extract buttons: (1,3) (2) etc.
        Pattern buttonRe = Pattern.compile("\\(([^)]*)\\)");
        Matcher buttonMatch = buttonRe.matcher(line);
        while (buttonMatch.find()) {
            String btnStr = buttonMatch.group(1).trim();
            if (btnStr.isEmpty()) {
                result.buttons.add(new ArrayList<>());
            } else {
                List<Integer> lights = new ArrayList<>();
                String[] parts = btnStr.split(",");
                for (String part : parts) {
                    try {
                        lights.add(Integer.parseInt(part.trim()));
                    } catch (NumberFormatException e) {
                        // Skip invalid numbers
                    }
                }
                result.buttons.add(lights);
            }
        }
        
        // Extract joltages: {3,5,4,7}
        Pattern joltageRe = Pattern.compile("\\{([^}]+)\\}");
        Matcher joltageMatch = joltageRe.matcher(line);
        if (joltageMatch.find()) {
            String joltageStr = joltageMatch.group(1);
            String[] parts = joltageStr.split(",");
            for (String part : parts) {
                try {
                    result.joltages.add(Integer.parseInt(part.trim()));
                } catch (NumberFormatException e) {
                    // Skip invalid numbers
                }
            }
        }
        
        return result;
    }
    
    static Integer gaussianEliminationGF2(List<List<Boolean>> matrix, List<Boolean> target) {
        int numButtons = matrix.size();
        int numLights = target.size();
        
        // Create augmented matrix [A | b]
        List<List<Boolean>> aug = new ArrayList<>();
        for (int i = 0; i < numLights; i++) {
            List<Boolean> row = new ArrayList<>();
            for (int j = 0; j < numButtons; j++) {
                if (j < matrix.size() && i < matrix.get(j).size()) {
                    row.add(matrix.get(j).get(i));
                } else {
                    row.add(false);
                }
            }
            row.add(target.get(i));
            aug.add(row);
        }
        
        // Gaussian elimination mod 2
        int pivotRow = 0;
        int pivotCol = 0;
        
        while (pivotRow < numLights && pivotCol < numButtons) {
            // Find pivot
            Integer pivotIdx = null;
            for (int i = pivotRow; i < numLights; i++) {
                if (aug.get(i).get(pivotCol)) {
                    pivotIdx = i;
                    break;
                }
            }
            
            if (pivotIdx == null) {
                pivotCol++;
                continue;
            }
            
            // Swap rows
            if (pivotIdx != pivotRow) {
                Collections.swap(aug, pivotRow, pivotIdx);
            }
            
            // Eliminate
            for (int i = pivotRow + 1; i < numLights; i++) {
                if (aug.get(i).get(pivotCol)) {
                    for (int j = 0; j <= numButtons; j++) {
                        boolean val = aug.get(i).get(j) ^ aug.get(pivotRow).get(j);
                        aug.get(i).set(j, val);
                    }
                }
            }
            
            pivotRow++;
            pivotCol++;
        }
        
        // Check for inconsistency (0 = 1)
        for (int i = pivotRow; i < numLights; i++) {
            if (aug.get(i).get(numButtons)) {
                boolean hasNonZero = false;
                for (int j = 0; j < numButtons; j++) {
                    if (aug.get(i).get(j)) {
                        hasNonZero = true;
                        break;
                    }
                }
                if (!hasNonZero) {
                    return null; // No solution
                }
            }
        }
        
        // Back substitution to find solution (set free variables to 0 to minimize)
        List<Boolean> solution = new ArrayList<>(Collections.nCopies(numButtons, false));
        Set<Integer> usedRows = new HashSet<>();
        
        // Process from bottom up
        for (int i = numLights - 1; i >= 0; i--) {
            // Find first non-zero column
            Integer pivotColIdx = null;
            for (int j = 0; j < numButtons; j++) {
                if (aug.get(i).get(j) && !usedRows.contains(j)) {
                    pivotColIdx = j;
                    usedRows.add(j);
                    break;
                }
            }
            
            if (pivotColIdx != null) {
                // Calculate value
                boolean val = aug.get(i).get(numButtons);
                for (int j = pivotColIdx + 1; j < numButtons; j++) {
                    if (aug.get(i).get(j) && solution.get(j)) {
                        val = !val;
                    }
                }
                solution.set(pivotColIdx, val);
            }
        }
        
        // Count number of presses
        int count = 0;
        for (boolean x : solution) {
            if (x) count++;
        }
        return count;
    }
    
    static Integer solvePart2ILP(List<List<Integer>> buttons, List<Integer> joltages) {
        int numButtons = buttons.size();
        int numLights = joltages.size();
        
        // Bounded search: max presses per button is max joltage
        int maxJoltage = joltages.stream().mapToInt(i -> i).max().orElse(0);
        
        return dfs(0, new ArrayList<>(Collections.nCopies(numLights, 0)), 0, null, buttons, joltages);
    }
    
    static Integer dfs(int buttonIdx, List<Integer> currentJoltages, int pressesSoFar, Integer best,
                       List<List<Integer>> buttons, List<Integer> joltages) {
        if (buttonIdx >= buttons.size()) {
            boolean allMatch = true;
            for (int i = 0; i < currentJoltages.size(); i++) {
                if (currentJoltages.get(i) != joltages.get(i)) {
                    allMatch = false;
                    break;
                }
            }
            if (allMatch) {
                return best == null ? pressesSoFar : Math.min(best, pressesSoFar);
            }
            return best;
        }
        
        // Pruning: if current presses already exceed best, skip
        if (best != null && pressesSoFar >= best) {
            return best;
        }
        
        // Calculate minimum additional presses needed (lower bound)
        int sumRemaining = 0;
        for (int i = 0; i < joltages.size(); i++) {
            int need = joltages.get(i) - currentJoltages.get(i);
            if (need > 0) sumRemaining += need;
        }
        if (sumRemaining > 0) {
            int maxLightsPerButton = 1;
            for (int i = buttonIdx; i < buttons.size(); i++) {
                if (buttons.get(i).size() > maxLightsPerButton) {
                    maxLightsPerButton = buttons.get(i).size();
                }
            }
            if (maxLightsPerButton > 0) {
                int minAdditional = (sumRemaining + maxLightsPerButton - 1) / maxLightsPerButton;
                if (best != null && pressesSoFar + minAdditional >= best) {
                    return best;
                }
            }
        }
        
        // Try 0 to maxJoltage presses of current button
        Integer bestResult = best;
        int maxJoltage = joltages.stream().mapToInt(i -> i).max().orElse(0);
        for (int presses = 0; presses <= maxJoltage; presses++) {
            if (best != null && pressesSoFar + presses >= best) {
                break;
            }
            
            List<Integer> newJoltages = new ArrayList<>(currentJoltages);
            for (int light : buttons.get(buttonIdx)) {
                if (light >= 0 && light < joltages.size()) {
                    newJoltages.set(light, newJoltages.get(light) + presses);
                }
            }
            
            // Check if any exceeds target (pruning)
            boolean exceeds = false;
            for (int i = 0; i < newJoltages.size(); i++) {
                if (newJoltages.get(i) > joltages.get(i)) {
                    exceeds = true;
                    break;
                }
            }
            if (exceeds) continue;
            
            Integer result = dfs(buttonIdx + 1, newJoltages, pressesSoFar + presses, bestResult, buttons, joltages);
            if (result != null) {
                bestResult = bestResult == null ? result : Math.min(bestResult, result);
            }
        }
        
        return bestResult;
    }
    
    public static String[] solve(String inputData) {
        String[] lines = inputData.trim().split("\n");
        if (lines.length == 0) {
            return new String[]{"0", "0"};
        }
        
        int part1Total = 0;
        int part2Total = 0;
        
        for (String line : lines) {
            if (line.trim().isEmpty()) continue;
            
            ParseResult parsed = parseLine(line);
            if (parsed.targetPattern.isEmpty()) continue;
            
            int numLights = parsed.targetPattern.size();
            
            // Part 1: GF(2) linear system
            // Build incidence matrix: matrix[i][j] = true if button i toggles light j
            List<List<Boolean>> buttonMatrix = new ArrayList<>();
            for (int i = 0; i < parsed.buttons.size(); i++) {
                List<Boolean> row = new ArrayList<>(Collections.nCopies(numLights, false));
                buttonMatrix.add(row);
            }
            for (int i = 0; i < parsed.buttons.size(); i++) {
                for (int light : parsed.buttons.get(i)) {
                    if (light >= 0 && light < numLights) {
                        buttonMatrix.get(i).set(light, true);
                    }
                }
            }
            
            // Target: all start OFF, need to toggle to match pattern
            List<Boolean> requiredToggles = new ArrayList<>(parsed.targetPattern);
            
            Integer result = gaussianEliminationGF2(buttonMatrix, requiredToggles);
            if (result != null) {
                part1Total += result;
            }
            
            // Part 2: Integer Linear Programming
            if (parsed.joltages.size() == numLights) {
                Integer result2 = solvePart2ILP(parsed.buttons, parsed.joltages);
                if (result2 != null) {
                    part2Total += result2;
                }
            }
        }
        
        return new String[]{String.valueOf(part1Total), String.valueOf(part2Total)};
    }
    
    public static void main(String[] args) throws IOException {
        Path inputPath = Paths.get("../data/input.txt");
        String data = Files.readString(inputPath);
        String[] results = solve(data);
        System.out.println("Part 1: " + results[0]);
        System.out.println("Part 2: " + results[1]);
    }
}

File: days/day_10/solutions/solution.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

typedef struct {
    int* target_pattern;
    int target_pattern_size;
    int** buttons;
    int* button_sizes;
    int num_buttons;
    int* joltages;
    int joltages_size;
} ParseResult;

void free_parse_result(ParseResult* pr) {
    if (pr->target_pattern) free(pr->target_pattern);
    if (pr->buttons) {
        for (int i = 0; i < pr->num_buttons; i++) {
            if (pr->buttons[i]) free(pr->buttons[i]);
        }
        free(pr->buttons);
    }
    if (pr->button_sizes) free(pr->button_sizes);
    if (pr->joltages) free(pr->joltages);
}

// Simple regex-like parsing without regex library
ParseResult parse_line(const char* line) {
    ParseResult result = {0};
    
    // Find pattern: [.##.]
    const char* pattern_start = strchr(line, '[');
    if (pattern_start) {
        const char* pattern_end = strchr(pattern_start + 1, ']');
        if (pattern_end) {
            int len = pattern_end - pattern_start - 1;
            result.target_pattern = (int*)malloc(len * sizeof(int));
            result.target_pattern_size = len;
            for (int i = 0; i < len; i++) {
                result.target_pattern[i] = (pattern_start[1 + i] == '#') ? 1 : 0;
            }
        }
    }
    
    // Extract buttons: (1,3) (2) etc.
    result.buttons = (int**)malloc(100 * sizeof(int*));
    result.button_sizes = (int*)malloc(100 * sizeof(int));
    result.num_buttons = 0;
    
    const char* pos = line;
    while ((pos = strchr(pos, '(')) != NULL) {
        const char* btn_start = pos + 1;
        const char* btn_end = strchr(btn_start, ')');
        if (!btn_end) break;
        
        int btn_len = btn_end - btn_start;
        if (btn_len > 0) {
            char* btn_str = (char*)malloc((btn_len + 1) * sizeof(char));
            strncpy(btn_str, btn_start, btn_len);
            btn_str[btn_len] = '\0';
            
            // Parse comma-separated numbers
            int* lights = (int*)malloc(100 * sizeof(int));
            int light_count = 0;
            char* token = strtok(btn_str, ",");
            while (token) {
                // Skip whitespace
                while (*token == ' ') token++;
                lights[light_count++] = atoi(token);
                token = strtok(NULL, ",");
            }
            
            result.buttons[result.num_buttons] = (int*)realloc(lights, light_count * sizeof(int));
            result.button_sizes[result.num_buttons] = light_count;
            result.num_buttons++;
            
            free(btn_str);
        } else {
            result.buttons[result.num_buttons] = NULL;
            result.button_sizes[result.num_buttons] = 0;
            result.num_buttons++;
        }
        pos = btn_end + 1;
    }
    
    // Extract joltages: {3,5,4,7}
    const char* joltage_start = strchr(line, '{');
    if (joltage_start) {
        const char* joltage_end = strchr(joltage_start + 1, '}');
        if (joltage_end) {
            int len = joltage_end - joltage_start - 1;
            char* joltage_str = (char*)malloc((len + 1) * sizeof(char));
            strncpy(joltage_str, joltage_start + 1, len);
            joltage_str[len] = '\0';
            
            result.joltages = (int*)malloc(100 * sizeof(int));
            result.joltages_size = 0;
            
            char* token = strtok(joltage_str, ",");
            while (token) {
                while (*token == ' ') token++;
                result.joltages[result.joltages_size++] = atoi(token);
                token = strtok(NULL, ",");
            }
            
            result.joltages = (int*)realloc(result.joltages, result.joltages_size * sizeof(int));
            free(joltage_str);
        }
    }
    
    return result;
}

int gaussian_elimination_gf2(int** matrix, int num_buttons, int num_lights, int* target, int* result_count) {
    // Create augmented matrix [A | b]
    int** aug = (int**)malloc(num_lights * sizeof(int*));
    for (int i = 0; i < num_lights; i++) {
        aug[i] = (int*)malloc((num_buttons + 1) * sizeof(int));
        for (int j = 0; j < num_buttons; j++) {
            if (j < num_buttons && i < num_lights && matrix[j]) {
                aug[i][j] = matrix[j][i];
            } else {
                aug[i][j] = 0;
            }
        }
        aug[i][num_buttons] = target[i];
    }
    
    // Gaussian elimination mod 2
    int pivot_row = 0;
    int pivot_col = 0;
    
    while (pivot_row < num_lights && pivot_col < num_buttons) {
        // Find pivot
        int pivot_idx = -1;
        for (int i = pivot_row; i < num_lights; i++) {
            if (aug[i][pivot_col]) {
                pivot_idx = i;
                break;
            }
        }
        
        if (pivot_idx == -1) {
            pivot_col++;
            continue;
        }
        
        // Swap rows
        if (pivot_idx != pivot_row) {
            int* temp = aug[pivot_row];
            aug[pivot_row] = aug[pivot_idx];
            aug[pivot_idx] = temp;
        }
        
        // Eliminate
        for (int i = pivot_row + 1; i < num_lights; i++) {
            if (aug[i][pivot_col]) {
                for (int j = 0; j <= num_buttons; j++) {
                    aug[i][j] ^= aug[pivot_row][j];
                }
            }
        }
        
        pivot_row++;
        pivot_col++;
    }
    
    // Check for inconsistency (0 = 1)
    for (int i = pivot_row; i < num_lights; i++) {
        if (aug[i][num_buttons]) {
            int has_non_zero = 0;
            for (int j = 0; j < num_buttons; j++) {
                if (aug[i][j]) {
                    has_non_zero = 1;
                    break;
                }
            }
            if (!has_non_zero) {
                // Free memory
                for (int i = 0; i < num_lights; i++) free(aug[i]);
                free(aug);
                return 0; // No solution
            }
        }
    }
    
    // Back substitution to find solution (set free variables to 0 to minimize)
    int* solution = (int*)calloc(num_buttons, sizeof(int));
    int* used_rows = (int*)calloc(num_buttons, sizeof(int));
    
    // Process from bottom up
    for (int i = num_lights - 1; i >= 0; i--) {
        // Find first non-zero column
        int pivot_col_idx = -1;
        for (int j = 0; j < num_buttons; j++) {
            if (aug[i][j] && !used_rows[j]) {
                pivot_col_idx = j;
                used_rows[j] = 1;
                break;
            }
        }
        
        if (pivot_col_idx != -1) {
            // Calculate value
            int val = aug[i][num_buttons];
            for (int j = pivot_col_idx + 1; j < num_buttons; j++) {
                if (aug[i][j] && solution[j]) {
                    val ^= 1;
                }
            }
            solution[pivot_col_idx] = val;
        }
    }
    
    // Count number of presses
    int count = 0;
    for (int i = 0; i < num_buttons; i++) {
        if (solution[i]) count++;
    }
    
    // Free memory
    free(solution);
    free(used_rows);
    for (int i = 0; i < num_lights; i++) free(aug[i]);
    free(aug);
    
    *result_count = count;
    return 1; // Solution found
}

int solve_part2_ilp(int** buttons, int* button_sizes, int num_buttons, int* joltages, int num_lights, int button_idx, int* current_joltages, int presses_so_far, int* best) {
    if (button_idx >= num_buttons) {
        int all_match = 1;
        for (int i = 0; i < num_lights; i++) {
            if (current_joltages[i] != joltages[i]) {
                all_match = 0;
                break;
            }
        }
        if (all_match) {
            if (*best == -1 || presses_so_far < *best) {
                *best = presses_so_far;
            }
            return *best;
        }
        return *best;
    }
    
    // Pruning: if current presses already exceed best, skip
    if (*best != -1 && presses_so_far >= *best) {
        return *best;
    }
    
    // Calculate minimum additional presses needed (lower bound)
    int sum_remaining = 0;
    for (int i = 0; i < num_lights; i++) {
        int need = joltages[i] - current_joltages[i];
        if (need > 0) sum_remaining += need;
    }
    if (sum_remaining > 0) {
        int max_lights_per_button = 1;
        for (int i = button_idx; i < num_buttons; i++) {
            if (button_sizes[i] > max_lights_per_button) {
                max_lights_per_button = button_sizes[i];
            }
        }
        if (max_lights_per_button > 0) {
            int min_additional = (sum_remaining + max_lights_per_button - 1) / max_lights_per_button;
            if (*best != -1 && presses_so_far + min_additional >= *best) {
                return *best;
            }
        }
    }
    
    // Find max joltage
    int max_joltage = 0;
    for (int i = 0; i < num_lights; i++) {
        if (joltages[i] > max_joltage) max_joltage = joltages[i];
    }
    
    // Try 0 to max_joltage presses of current button
    int best_result = *best;
    for (int presses = 0; presses <= max_joltage; presses++) {
        if (*best != -1 && presses_so_far + presses >= *best) {
            break;
        }
        
        int* new_joltages = (int*)malloc(num_lights * sizeof(int));
        memcpy(new_joltages, current_joltages, num_lights * sizeof(int));
        
        for (int k = 0; k < button_sizes[button_idx]; k++) {
            int light = buttons[button_idx][k];
            if (light >= 0 && light < num_lights) {
                new_joltages[light] += presses;
            }
        }
        
        // Check if any exceeds target (pruning)
        int exceeds = 0;
        for (int i = 0; i < num_lights; i++) {
            if (new_joltages[i] > joltages[i]) {
                exceeds = 1;
                break;
            }
        }
        
        if (!exceeds) {
            int result = solve_part2_ilp(buttons, button_sizes, num_buttons, joltages, num_lights, button_idx + 1, new_joltages, presses_so_far + presses, best);
            if (result != -1 && (best_result == -1 || result < best_result)) {
                best_result = result;
                *best = result;
            }
        }
        
        free(new_joltages);
    }
    
    return best_result;
}

void solve(const char* input_data, long long* part1_result, long long* part2_result) {
    // Split into lines
    char** lines = NULL;
    int line_count = 0;
    int capacity = 1000;
    lines = (char**)malloc(capacity * sizeof(char*));
    
    const char* pos = input_data;
    while (*pos != '\0') {
        const char* line_start = pos;
        while (*pos != '\n' && *pos != '\0') pos++;
        
        if (line_count >= capacity) {
            capacity *= 2;
            lines = (char**)realloc(lines, capacity * sizeof(char*));
        }
        
        int len = pos - line_start;
        if (len > 0) {
            lines[line_count] = (char*)malloc((len + 1) * sizeof(char));
            strncpy(lines[line_count], line_start, len);
            lines[line_count][len] = '\0';
        } else {
            lines[line_count] = NULL;
        }
        line_count++;
        
        if (*pos == '\n') pos++;
    }
    
    *part1_result = 0;
    *part2_result = 0;
    
    for (int i = 0; i < line_count; i++) {
        if (!lines[i] || strlen(lines[i]) == 0) continue;
        
        ParseResult parsed = parse_line(lines[i]);
        if (parsed.target_pattern_size == 0) {
            free_parse_result(&parsed);
            continue;
        }
        
        int num_lights = parsed.target_pattern_size;
        
        // Part 1: GF(2) linear system
        // Build incidence matrix: matrix[i][j] = 1 if button i toggles light j
        int** button_matrix = (int**)malloc(parsed.num_buttons * sizeof(int*));
        for (int j = 0; j < parsed.num_buttons; j++) {
            button_matrix[j] = (int*)calloc(num_lights, sizeof(int));
            for (int k = 0; k < parsed.button_sizes[j]; k++) {
                int light = parsed.buttons[j][k];
                if (light >= 0 && light < num_lights) {
                    button_matrix[j][light] = 1;
                }
            }
        }
        
        // Required toggles: target_pattern
        int result_count = 0;
        if (gaussian_elimination_gf2(button_matrix, parsed.num_buttons, num_lights, parsed.target_pattern, &result_count)) {
            *part1_result += result_count;
        }
        
        // Free button matrix
        for (int j = 0; j < parsed.num_buttons; j++) {
            free(button_matrix[j]);
        }
        free(button_matrix);
        
        // Part 2: Integer Linear Programming
        if (parsed.joltages_size == num_lights) {
            int* initial_joltages = (int*)calloc(num_lights, sizeof(int));
            int best = -1;
            int result = solve_part2_ilp(parsed.buttons, parsed.button_sizes, parsed.num_buttons, parsed.joltages, num_lights, 0, initial_joltages, 0, &best);
            if (result != -1) {
                *part2_result += result;
            }
            free(initial_joltages);
        }
        
        free_parse_result(&parsed);
    }
    
    // Free lines
    for (int i = 0; i < line_count; i++) {
        if (lines[i]) free(lines[i]);
    }
    free(lines);
}

int main() {
    FILE* file = fopen("../data/input.txt", "r");
    if (file == NULL) {
        printf("Error opening file\n");
        return 1;
    }
    
    fseek(file, 0, SEEK_END);
    long size = ftell(file);
    fseek(file, 0, SEEK_SET);
    
    char* content = (char*)malloc((size + 1) * sizeof(char));
    fread(content, 1, size, file);
    content[size] = '\0';
    fclose(file);
    
    long long part1, part2;
    solve(content, &part1, &part2);
    
    printf("Part 1: %lld\n", part1);
    printf("Part 2: %lld\n", part2);
    
    free(content);
    return 0;
}
